### ISSUE TITLE
Unification of class names across languages

### ISSUE THREAD

**Issue by isagalaev** [isagalaev/highlight.js#348]
## Problem

>  One of the early design principles for highlight.js was having just a few common class names in order to have universal styles that would work for any language. Unfortunately this principle wasn't strongly enforced. We now have language-specific classes, language-specific style rules and whole language-specific styles. This is a maintenance nightmare as the number of unique conditions that should be visually tested is a production of the number of unique language features and the number of styles. And since this is an impossible amount of work, we usually test only a small subset of languages and styles and rely on pure luck for the rest which is a majority. ([source][])

## Principles

- Class names should have stronger semantics. A "variable" in PHP is hardly worth highlighting at all while a "variable" in Prolog is essential for understanding the code and is as important as, say, function definitions in other languages. This means that we'll have to invent a separate name for every case that has to be treated differently in different languages.

- The set of available class names should be rich enough for all languages and adding a new one after we convert all we currently have should require a serious reason.

- The class reference becomes a flat list of available class names with explanations of their meaning.

- All styles become language-agnostic. We may even drop generating class names for languages altogether.

## Process

There is a a long lived branch where the conversion is slowly happening: https://github.com/isagalaev/highlight.js/tree/new-styles.

It should eliminate all language names from CSS files. Every current class name should also be evaluated and many will be either dropped or change their semantics. At every such change (removal of a language specific rule or redefinition of a class name) we should make sure that *all* styles look reasonable. This means that we'll definitely loose similarity with other highlighter engines emulated by various styles, but that's an inevitable implication of the whole task.

During the whole conversion process we will have to intelligently port new languages and styles into the new rules. Possibly encouraging contributors to do that :-)

[source]: http://softwaremaniacs.org/blog/2014/07/26/highlight-js-what-next/en/

## Status

- [x] 1c
- [x] accesslog
- [x] actionscript
- [x] apache
- [x] applescript
- [x] asciidoc
- [x] armasm
- [x] aspectj
- [x] autohotkey
- [x] autoit
- [x] avrasm
- [x] axapta
- [x] bash
- [x] brainfuck
- [x] cal
- [x] capnproto
- [x] ceylon
- [x] clojure
- [x] clojure-repl
- [x] cmake
- [x] coffeescript
- [x] cpp
- [x] crmsh
- [x] crystal
- [x] cs
- [x] css
- [x] d
- [x] markdown
- [x] dart
- [x] delphi
- [x] diff
- [x] django
- [x] dns
- [x] dockerfile
- [x] dos
- [x] dust
- [x] elixir
- [x] elm
- [x] erb
- [x] erlang-repl
- [x] erlang
- [x] fix
- [x] fortran
- [x] fsharp
- [x] gams
- [x] gcode
- [x] gherkin
- [x] glsl
- [x] go
- [x] golo
- [x] gradle
- [x] groovy
- [x] haml
- [x] handlebars
- [x] haskell
- [x] haxe
- [x] http
- [x] inform7
- [x] irpf90
- [x] ini
- [x] java
- [x] javascript
- [x] json
- [x] julia
- [x] kotlin
- [x] lasso
- [x] less
- [x] lisp
- [x] livecodeserver
- [x] livescript
- [x] lua
- [x] makefile
- [x] mathematica
- [x] matlab
- [x] mel
- [x] mercury
- [x] mizar
- [x] mojolicious
- [x] monkey
- [x] nginx
- [x] nimrod
- [x] nix
- [x] nsis
- [x] objectivec
- [x] ocaml
- [x] openscad
- [x] oxygene
- [x] parser3
- [x] perl
- [x] pf
- [x] php
- [x] powershell
- [x] processing
- [x] profile
- [x] prolog
- [x] protobuf
- [x] puppet
- [x] python
- [x] q
- [x] r
- [x] rib
- [x] roboconf
- [x] rsl
- [x] ruby
- [x] ruleslanguage
- [x] rust
- [x] scala
- [x] scheme
- [x] scilab
- [x] scss
- [x] smali
- [x] smalltalk
- [x] sml
- [x] sqf
- [x] sql
- [x] stata
- [x] step21
- [x] stylus
- [x] swift
- [x] tcl
- [x] tex
- [x] thrift
- [x] tp
- [x] twig
- [x] typescript
- [x] vala
- [x] vbnet
- [x] vbscript
- [x] vbscript-html
- [x] verilog
- [x] vhdl
- [x] vim
- [x] x86asm
- [x] xl
- [x] xml
- [x] xquery
- [x] zephir

**Comment by isagalaev** [comment]
It's begun: http://softwaremaniacs.org/blog/2015/05/06/styles-unification-first-results/en/

Everyone feeling brave is welcome to join the work on the new-styles branch! The way I do it is roughly this:

- Take a language file at random.
- For every class names being used make sure it doesn't feel wrong (like "tag" for section headers in Apache) and it is defined in the class reference. If there's no suitable name there already, invent and document a new one. This is a sensitive part of the job with no formal criteria, so I'd be happy to discuss that.
- Go through all the styles and make sure that the visual result for this particular language is good. The painful part is that many styles are quite inconsistent and you have to second-guess the author as to what things were supposed to mean. For example there's new class ".hljs-meta" that is supposed to replace all the old preprocessor directives, processing instructions, shell prompts, etc. The problem is that many styles define different colors for those and you have to try all of them to find that looks fine.
- Along the way, remove all language-specific and dead class names from this iteration from styles.

**Comment by non** [comment]
Just my 2¢ -- this is a great plan!

I initially started working on the Scala mode due to the use of highlight.js in Deckset. However, even after my changes were released, I was disappointed that Scala code still didn't highlight well in many (most?) styles, including the one Deckset seemed to be using.

I started doing what you suggested for Scala. Once I have something that seems good, shall I submit a PR against your branch?

**Comment by isagalaev** [comment]
@non yes, please do! I suggest doing it this way:

- update the language definition and the class reference (+ probably any one style as an example)
- submit the PR so we can discuss and settle the semantics of the classes
- then go ahead and update all the other styles

**Comment by non** [comment]
Sounds good.

**Comment by isagalaev** [comment]
@non just ported the rest of the styles with Scala changes. Check!

**Comment by non** [comment]
Thanks! :tophat:

**Comment by Sannis** [comment]
I'll push some languages to #911.

**Comment by Sannis** [comment]
Zephir missed in list.

**Comment by isagalaev** [comment]
Ah, I knew I was gonna miss something :-) Thanks!

**Comment by isagalaev** [comment]
Okay… Upon re-writing the style guide this looked pretty much ready and did burn my hands a lot, so I decided to merge it into master right away. I was thinking about publishing a separate beta demo page from the branch but my prior experience show that nobody ever wants to test beta versions of small libraries :-) I believe we've got enough time before the next release date to iron out possible quirks (see #967).

P.S. Woo-hoo!
